contracts/0.8.25/vaults/VaultFactory.sol

```diff
@@ -6,205 +6,111 @@ pragma solidity 0.8.25;
+     * @param _defaultAdmin The address of the default admin of the Dashboard
+     * @param _nodeOperator The address of the node operator of the StakingVault
+     * @param _nodeOperatorManager The address of the node operator manager in the Dashboard
+     * @param _nodeOperatorFeeBP The node operator fee in basis points
+     * @param _confirmExpiry The confirmation expiry in seconds
+     * @param _roleAssignments The optional role assignments to be made
+     * @param _extraParams The extra params for the StakingVault
+     */
+    function createVaultWithDashboard(
+        address _defaultAdmin,
+        address _nodeOperator,
+        address _nodeOperatorManager,
+        uint256 _nodeOperatorFeeBP,
+        uint256 _confirmExpiry,
+        Permissions.RoleAssignment[] calldata _roleAssignments,
+        bytes calldata _extraParams
+    ) external returns (IStakingVault vault, Dashboard dashboard) {
+        vault = IStakingVault(address(new BeaconProxy(BEACON, "")));
+
         bytes memory immutableArgs = abi.encode(vault);
-        delegation = Delegation(payable(Clones.cloneWithImmutableArgs(DELEGATION_IMPL, immutableArgs)));
+        dashboard = Dashboard(payable(Clones.cloneWithImmutableArgs(DASHBOARD_IMPL, immutableArgs)));
 
         // initialize StakingVault
         vault.initialize(
             address(this),
-            _delegationConfig.nodeOperatorManager,
+            _nodeOperator,
             ILidoLocator(LIDO_LOCATOR).predepositGuarantee(),
-            _stakingVaultInitializerExtraParams
+            _extraParams
         );
 
         vault.authorizeLidoVaultHub();
 
         // transfer ownership of the vault back to the delegation
-        OwnableUpgradeable(address(vault)).transferOwnership(address(delegation));
-
-        // initialize Delegation
-        delegation.initialize(address(this), _delegationConfig.confirmExpiry);
+        OwnableUpgradeable(address(vault)).transferOwnership(address(dashboard));
 
-        // setup roles from config
-        // basic permissions to the staking vault
-        delegation.grantRole(delegation.DEFAULT_ADMIN_ROLE(), _delegationConfig.defaultAdmin);
-        delegation.grantRole(delegation.NODE_OPERATOR_MANAGER_ROLE(), _delegationConfig.nodeOperatorManager);
-        delegation.grantRole(delegation.ASSET_RECOVERY_ROLE(), _delegationConfig.assetRecoverer);
+        // If there are extra role assignments to be made,
+        // we initialize the dashboard with the VaultFactory as the default admin,
+        // grant the roles and revoke the VaultFactory's admin role.
+        // Otherwise, we initialize the dashboard with the default admin.
+        if (_roleAssignments.length > 0) {
+            dashboard.initialize(address(this), _nodeOperatorManager, _nodeOperatorFeeBP, _confirmExpiry);
+            // will revert if any role is not controlled by the default admin
+            dashboard.grantRoles(_roleAssignments);
```

Comment:
integrate initial role assignments in initialize to avoid role switching and get rid of if-branch here?